plugin/pkg/admission/initialresources/admission.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialresources
+
+import (
+	"flag"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/resource"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+var (
+	source     = flag.String("ir-data-source", "influxdb", "Data source used by InitialResources. Supported options: influxdb.")
+	percentile = flag.Int64("ir-percentile", 90, "Which percentile of samples should InitialResources use when estimating resources. For experiment purposes.")
+)
+
+const (
+	week  = 7 * 24 * time.Hour
+	month = 30 * 24 * time.Hour
+)
+
+// WARNING: this feature is experimental and will definitely change.
+func init() {
+	admission.RegisterPlugin("InitialResources", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		s, err := newDataSource(*source)
+		if err != nil {
+			return nil, err
+		}
+		return newInitialResources(s), nil
+	})
+}
+
+type initialResources struct {
+	*admission.Handler
+	source dataSource
+}
+
+func newInitialResources(source dataSource) admission.Interface {
+	return &initialResources{
+		Handler: admission.NewHandler(admission.Create),
+		source:  source,
+	}
+}
+
+func (ir initialResources) Admit(a admission.Attributes) (err error) {
+	// Ignore all calls to subresources or resources other than pods.
+	if a.GetSubresource() != "" || a.GetResource() != string(api.ResourcePods) {
+		return nil
+	}
+	pod, ok := a.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	ir.estimateResourcesIfNotSet(pod)
+	return nil
+}
+
+func (ir initialResources) estimateResourcesIfNotSet(pod *api.Pod) {
```

Comment:
Please add a comment that this method has a side effect and sets the estimates inside of the given pod.
Nit: the behavior would be cleaner if you actually returned a valid and verified resources and it would be up to the method caller how and where to set it.